2단계 - 부하 테스트 리뷰 요청

yomni opened the pull request:
안녕하세요!
성능 테스트에 대한 학습이 부족하여,
이번에도 학습 후 미션 진행하였습니다.

성능 테스트 라는 것이 참 어려운 점이,
_**논리적인 목표를 설정하기 때문에, 테스트 결과 값을 100% 신뢰할 수는 없다.**_ 라는 점인것같습니다.

여러모로 다른 시스템을 사용하면서 사용자 관점에서 성능이 느리면 불편한 기색을 내비췄던 제가 부끄러워지네요..🫠
(예약 시스템 등을 이용할 때...)

따라서 일반적인 B2C 서비스를 운영할 때, 이런 지표들을 계속해서  
추적 & 관리 해야 될 필요성과어려움을 동시에 느끼는 미션이었습니다!

리뷰 미리 감사드립니다!

오늘부터 날이 추워진다는데 감기 조심하시길 바랍니다🥶

yomni committed:
feat: step2 부하 테스트 내용 작성

yomni committed:
style: 개행문자 추가

limwoobin commented on README.md:
```diff
@@ -162,19 +224,488 @@ A) 정답은 없다 ; 서비스 특성마다 다른 기준을 가지고 개선
   - 따라서, 안내되어 있는 내용 대로 gzip 압축등의 방법을 통해 성능예산 목표를 달성할 수 있도록 개선해 볼 수 있을 것 같습니다.
 
+#### 1단계 피드백
+- cdn 서버 적용으로도 성능개선 여지가 있음
+
+> CDN(Content Delivery Network) 서버란?  
+> 한마디로 말하면, **캐시 서버의 일종**이라고 할 수 있다.  
+> 최근 콘텐츠는 변동될 가능성이 매우 적다(netflex의 영상, 게임의 리소스 등등..)  
+> 이런 콘텐츠는 캐싱하여 Content 요청시에 CDN 서비스 등을 통해 응답하는 것이  
+> 성능적인 측면에서 훨씬 이점이 많다.
+
```
👍👍👍

limwoobin commented on README.md:
```diff
@@ -82,11 +82,73 @@ A) 정답은 없다 ; 서비스 특성마다 다른 기준을 가지고 개선
+  - 극한 상황에서의 동작 테스트
+  - 장기간 부하 발생에 대한 한계치 확인 및 기능 정상 동작 여부 확인
+  - 최대 사용자 또는 최대 처리량을 확인
+  - 스트레스 테스트 이후 시스템이 수동 개입 없이 복구되는 지 확인
```
굉장히 잘 정리해주셨네요!
저도 정말 많이 배우게되네요 ㅎㅎ

limwoobin commented on README.md:
```diff
@@ -82,11 +82,73 @@ A) 정답은 없다 ; 서비스 특성마다 다른 기준을 가지고 개선
+>   - TPS 는 `Scale Out : 분산처리` 혹은 `Scale up : 리소스 스펙업`을 통해 증가 가능
+>   - 단순히 응답시간을 기준으로 종료시키지 말고, TPS 나 DB Connection, CPU 등을 종합적으로 확인하고 중단해야 함
+> - Performance vs Scalability  
+> ![scaling.png](images%2Fstep2%2Fscaling.png)
```
파일경로가 깨진건지 잘못설정되어있네요ㅠ
이미지는 확인했습니다!

limwoobin commented on README.md:
```diff
@@ -162,19 +224,488 @@ A) 정답은 없다 ; 서비스 특성마다 다른 기준을 가지고 개선
+#### 부하테스트 전제조건은 어느정도로 설정하셨나요
+##### 대상 시스템 범위 선정
+
+지하철 노선도 앱의 주요 기능은 바로 `노선 조회` 입니다.
+따라서 타겟 시스템 범위는 다음과 같이 설정하였습니다.
+- `메인 페이지` : 노선도 시스템 최초 접근 화면
+- `노선 조회` : 데이터를 조회하는 데 여러 데이터를 참조하는 페이지
```
타겟시스템범위에 대해 잘 설명해주셧네요 ㅎㅎ
경로검색 페이지 접근도 추가하는것도 괜찮아보입니다!

limwoobin commented on k6scripts/load_path.js:
```diff
@@ -0,0 +1,28 @@
+    check(pathRes, {
+        'is success': (r) => r.status === 200,
+    });
+}
```
메인페이지 접근, 경로조회를 따로따로 테스트하신걸까요??
유저플로우를 보면 `메인접근 > 경로조회페이지접근 > 경로조회` 와 같은 흐름으로 진행될것으로 예상되는데
한번에 위 플로우를 모두 테스트하는건 어떨까요??

yomni replied:
아하 한 스크립트 내에서 플로우를 지정해서 테스트할 수 있는걸 몰랐습니다!
학습 후 모두 테스트해보는것으로 변경해서 진행해보겠습니다!

limwoobin commented on k6scripts/load_main.js:
```diff
@@ -0,0 +1,24 @@
+        {duration: '10s', target: 23},
+        {duration: '300s', target: 64},
+        {duration: '10s', target: 0},
+    ],
```
<img width="984" alt="image" src="https://user-images.githubusercontent.com/28802545/207550147-d789c871-4785-4fc4-aa95-a5aedd3088a3.png">
테스트 시간에 대해 한번 고민해보는건 어떠실까요?

yomni replied:
Load Test의 경우 30분 ~ 2시간 사이로 테스트 시행해보겠습니다!

limwoobin commented on k6scripts/load_main.js:
```diff
@@ -0,0 +1,24 @@
+
+export let options = {
+    threshold: {
+        http_req_duration: ['p(99)<1000'],
```
<img width="611" alt="image" src="https://user-images.githubusercontent.com/28802545/207550446-685d79ab-a754-4e8e-a232-3c08c6927946.png">

load, smoke 테스트에서 1000은 좀 긴거같기도 하네요!
stress 테스트에서도 살짝 낮춰보아도 괜찮을것같습니다!

yomni replied:
smoke는 정상적인 구동 여부를 확인하는 테스트로 이해했습니다.
따라서 1000으로 유지하고,

load, stress는 100이하로 낮춰보겠습니다!

limwoobin replied:
넵 말씀하신 방향 좋습니다 👍

limwoobin commented:
그라파나 결과, k6 결과 이미지가 누락된것같은데 같이 첨부 부탁드립니다!

limwoobin reviewed (changes requested):
안녕하세요 용민님
2단계 미션도 잘 해주셨네요 ㅎㅎ

지난번도 이번에도 사전정리를 정말 잘 해주셔서 
저도 리뷰하면서 많이 배우고있습니다
피드백에 대한 회고도 작성해주신것도 굉장히 좋네요 👍 

몇가지 코멘트를 남겨드렸는데 확인해주시고 다시 요청부탁드립니다!

yomni committed:
refactor: 2단계 피드백 적용

- 경로 깨진 이미지 파일 확인
- 스크립트 한개 내에서 유저 플로우를 정의하여 여러 화면/기능에 대한 check
- Load Test 의 테스트 시간을 30분 ~ 2시간 사이로 테스트(1시간)
- Load, Stress 테스트의 threshold 시간을 낮춰볼 것

limwoobin commented on k6scripts/stress.js:
```diff
@@ -0,0 +1,42 @@
+
+export let options = {
+    threshold: {
+        http_req_duration: ['p(99)<100'],
```
stress테스트에 대해서는 duration을 조금 더 길게 가져가셔도 좋지 않을까 싶은데 
어떻게 생각하시나요??
현재 설정하신 duration으로는 지표가 어떨지 궁금하기도 하네요 ㅎㅎ

limwoobin commented:
그라파나 테스트 결과이미지가 하나만 있는걸로 보여집니다.
smoke, load, stress 세가지 경우에 대해 이미지를 첨부해주시면 좋을것같아요!
다음 미션진행하며 올려주시면 좋겟습니다 ㅎㅎ

limwoobin reviewed (approved):
안녕하세요 용민님
2단계 미션하느라 고생많으셨습니다!

몇가지 코멘트를 남겨드렸으니 다음미션 진행하며
같이 확인해주시면 좋을듯 합니다!
다음 미션도 화이팅입니다!
